File: notification-domain/src/main/java/com/example/notification/domain/template/TemplateKey.java
package com.example.notification.domain.template;

import java.util.Objects;

/**
 * 템플릿 식별자. 운영자가 등록한 템플릿을 use case 가 참조할 때 쓰는 안정 키.
 *
 * <p>예: {@code order.shipped.v1}, {@code auth.otp.v2}.
 *
 * <p>관례: {@code <도메인>.<이벤트>.v<버전>}. 버전이 바뀌면 새 row 로 등록.
 */
public final class TemplateKey {

    private final String value;

    public TemplateKey(String value) {
        Objects.requireNonNull(value, "templateKey must not be null");
        String trimmed = value.trim();
        if (!trimmed.matches("^[a-z][a-z0-9_]*(\\.[a-z][a-z0-9_]*)+$")) {
            throw new IllegalArgumentException(
                    "templateKey must match `^[a-z][a-z0-9_]*(\\.[a-z][a-z0-9_]*)+$`: " + value);
        }
        this.value = trimmed;
    }

    public String value() {
        return value;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof TemplateKey other)) return false;
        return value.equals(other.value);
    }

    @Override
    public int hashCode() {
        return value.hashCode();
    }

    @Override
    public String toString() {
        return value;
    }
}

File: notification-domain/src/main/kotlin/com/example/notification/domain/template/TemplateKey.kt
package com.example.notification.domain.template

/**
 * 템플릿 식별자. 운영자가 등록한 템플릿을 use case 가 참조할 때 쓰는 안정 키.
 *
 * 예: `order.shipped.v1`, `auth.otp.v2`.
 *
 * 관례: `<도메인>.<이벤트>.v<버전>`. 버전이 바뀌면 새 row 로 등록.
 *
 * 생성자에서 trim 한 정규화 값을 보존하므로 equals/hashCode 는 직접 구현.
 */
class TemplateKey(value: String) {

    @get:JvmName("value")
    val value: String = value.trim()

    init {
        require(this.value.matches(KEY_PATTERN)) {
            "templateKey must match `^[a-z][a-z0-9_]*(\\.[a-z][a-z0-9_]*)+$`: $value"
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is TemplateKey) return false
        return value == other.value
    }

    override fun hashCode(): Int = value.hashCode()

    override fun toString(): String = value

    companion object {
        private val KEY_PATTERN = Regex("^[a-z][a-z0-9_]*(\\.[a-z][a-z0-9_]*)+$")
    }
}
